// @nrwl/react/src/utils/create-ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractTsConfigBase = exports.createTsConfig = void 0;
const shared = require("@nrwl/workspace/src/utils/create-ts-config");
const json_1 = require("nx/src/generators/utils/json");
function createTsConfig(host, projectRoot, type, options, relativePathToRootTsConfig) {
    const json = {
        compilerOptions: {
            jsx: 'react-jsx',
            allowJs: false,
            esModuleInterop: false,
            allowSyntheticDefaultImports: true,
            strict: options.strict,
        },
        files: [],
        include: [],
        references: [
            {
                path: type === 'app' ? './tsconfig.app.json' : './tsconfig.lib.json',
            },
        ],
    };
    if (options.style === '@emotion/styled') {
        json.compilerOptions.jsxImportSource = '@emotion/react';
    }
    if (options.bundler === 'vite') {
        json.compilerOptions.types =
            options.unitTestRunner === 'vitest'
                ? ['vite/client', 'vitest']
                : ['vite/client'];
    }
    // inline tsconfig.base.json into the project
    if (options.rootProject) {
        json.compileOnSave = false;
        json.compilerOptions = Object.assign(Object.assign({}, shared.tsConfigBaseOptions), json.compilerOptions);
        json.exclude = ['node_modules', 'tmp'];
    }
    else {
        json.extends = relativePathToRootTsConfig;
    }
    (0, json_1.writeJson)(host, `${projectRoot}/tsconfig.json`, json);
}
exports.createTsConfig = createTsConfig;
function extractTsConfigBase(host) {
    shared.extractTsConfigBase(host);
    if (host.exists('vite.config.ts')) {
        const vite = host.read('vite.config.ts').toString();
        host.write('vite.config.ts', vite.replace(`projects: []`, `projects: ['tsconfig.base.json']`));
    }
}
exports.extractTsConfigBase = extractTsConfigBase;
