// @nrwl/nest/src/generators/application/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toNodeApplicationGeneratorOptions = exports.normalizeOptions = void 0;
const devkit_1 = require("@nrwl/devkit");
const devkit_2 = require("@nrwl/devkit");
const linter_1 = require("@nrwl/linter");
function normalizeOptions(tree, options) {
    var _a, _b;
    const { layoutDirectory, projectDirectory } = (0, devkit_1.extractLayoutDirectory)(options.directory);
    const appDirectory = projectDirectory
        ? `${(0, devkit_2.names)(projectDirectory).fileName}/${(0, devkit_2.names)(options.name).fileName}`
        : (0, devkit_2.names)(options.name).fileName;
    const appProjectRoot = (0, devkit_2.joinPathFragments)(layoutDirectory !== null && layoutDirectory !== void 0 ? layoutDirectory : (0, devkit_2.getWorkspaceLayout)(tree).appsDir, appDirectory);
    return Object.assign(Object.assign({}, options), { appProjectRoot, linter: (_a = options.linter) !== null && _a !== void 0 ? _a : linter_1.Linter.EsLint, unitTestRunner: (_b = options.unitTestRunner) !== null && _b !== void 0 ? _b : 'jest' });
}
exports.normalizeOptions = normalizeOptions;
function toNodeApplicationGeneratorOptions(options) {
    return {
        name: options.name,
        directory: options.directory,
        frontendProject: options.frontendProject,
        linter: options.linter,
        skipFormat: true,
        skipPackageJson: options.skipPackageJson,
        standaloneConfig: options.standaloneConfig,
        tags: options.tags,
        unitTestRunner: options.unitTestRunner,
        setParserOptionsProject: options.setParserOptionsProject,
        bundler: 'webpack', // Some features require webpack plugins such as TS transformers
    };
}
exports.toNodeApplicationGeneratorOptions = toNodeApplicationGeneratorOptions;
